src/main/scala/scalaz/stream/Process.scala

```diff
@@ -1192,93 +1207,261 @@ object Process extends ProcessInstances {
+
+                                case None => ???        // wtf?! (apparently we were called twice with None; bug in completeInterruptibly)
+                              }
+                            }
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+
+                /*
+                 * Start the task. per the `completeInterruptibly` invariants, the callback will be invoked exactly once
+                 * unless interrupted in the final computation step, in which case it will be invoked twice: once with
+                 * the interrupt signal and once with the final computed result (this case is detected by the `PostStep`)
+                 * extractor.  Under all other circumstances, including interrupts, exceptions and natural completion, the
+                 * callback will be invoked exactly once.
+                 */
+                lazy val interrupt: () => Unit = completeInterruptibly((checkInterrupt(interrupt) >> req).get)(handle(
+                  preStep = { cause =>
+                    // interrupted; now drain
+                    (Try(rcv(-\/(cause)).run) +: cont).drain.run runAsync {
+                      case -\/(t) => S { cb(-\/(Error(t) causedBy cause)) }
+                      case \/-(_) => S { cb(-\/(cause)) }
+                    }
+                  },
+
+                  midStep = { cause =>
+                    // interrupted; now drain
+                    (Try(rcv(-\/(cause)).run) +: cont).drain.run runAsync {
+                      case -\/(t) => S { cb(-\/(Error(t) causedBy cause)) }
+                      case \/-(_) => S { cb(-\/(cause)) }
+                    }
+                  },
+
+                  postStep = { (inner, cause) =>
+                    inner.kill.run runAsync { _ => () }
+                  },
+
+                  exceptional = { t =>
+                    // we got an exception (not an interrupt!) and we need to drain everything
+                    go(Try(rcv(-\/(Error(t))).run) +: cont) runAsync { _ => () }
+                  },
+
+                  completed = { continuation =>
+                    go(continuation) runAsync { _ => () }
+                  }
+                ))
+
+                interrupt     // please don't delete this!  highly mutable code within
+
+                // interrupts the current step (may be a recursive child!) and sets `interrupted`
+                def referencedInterrupt(cause: EarlyCause): Unit = {
+                  interrupted.get() match {
+                    case ptr @ -\/(int) => {
+                      if (interrupted.compareAndSet(ptr, \/-(cause))) {
+                        int()
+                      } else {
+                        referencedInterrupt(cause)
+                      }
+                    }
+
+                    case \/-(_) => ()    // ?????
```

Comment:
I'll elaborate on them.  Basically, that case can only occur if we are interrupted multiple times with different causes.  We always drop the later causes on the floor in that case.